third_party/amd/lib/TritonAMDGPUTransforms/CoalesceAsyncCopy.cpp

```diff
@@ -0,0 +1,193 @@
+    int threadsPerWarp = ttg::TritonGPUDialect::getThreadsPerWarp(mod);
+    auto newBlockEnc = BlockedEncodingAttr::get(
+        copyOp.getContext(), srcTy.getShape(), contigPerThread,
+        blockedEnc.getOrder(), numWarps, threadsPerWarp,
+        blockedEnc.getCTALayout());
+
+    // Convert layout of src, mask and other to new encoding
+    auto convertLayout = [&rewriter](auto loc, Value old, auto newEnc) {
+      auto oldTy = cast<RankedTensorType>(old.getType());
+      RankedTensorType newSrcTy = RankedTensorType::get(
+          oldTy.getShape(), oldTy.getElementType(), newEnc);
+      return rewriter.create<ttg::ConvertLayoutOp>(loc, newSrcTy, old);
+    };
+
+    auto loc = copyOp->getLoc();
+    Value cvtSrc = convertLayout(loc, src, newBlockEnc);
+
+    if (mask)
+      mask = convertLayout(loc, mask, newBlockEnc);
+    if (other)
+      other = convertLayout(loc, other, newBlockEnc);
+
+    rewriter.modifyOpInPlace(copyOp, [&]() {
+      copyOp.getSrcMutable().assign(cvtSrc);
+      if (mask)
+        copyOp.getMaskMutable().assign(mask);
+      if (other)
+        copyOp.getOtherMutable().assign(other);
+    });
+    return success();
+  }
+
+private:
+  const triton::AMD::TargetInfo &targetInfo;
+  DenseMap<ttg::AsyncCopyGlobalToLocalOp, unsigned> asyncCopyContiguity;
+};
+
+class TritonAMDGPUCoalesceAsyncCopyPass
+    : public TritonAMDGPUCoalesceAsyncCopyBase<
+          TritonAMDGPUCoalesceAsyncCopyPass> {
+public:
+  TritonAMDGPUCoalesceAsyncCopyPass(StringRef archGenName) {
+    this->archGenerationName = archGenName.str();
+  }
+
+  void runOnOperation() override {
+    ModuleOp m = getOperation();
+    MLIRContext *context = &getContext();
+
+    triton::AMD::TargetInfo targetInfo(archGenerationName);
+
+    mlir::RewritePatternSet patterns(context);
+
+    switch (targetInfo.getISAFamily()) {
+    case triton::AMD::ISAFamily::CDNA1:
+    case triton::AMD::ISAFamily::CDNA2:
+    case triton::AMD::ISAFamily::CDNA3:
+    case triton::AMD::ISAFamily::CDNA4: {
+      // Precompute the contiguity of all AsyncCopy ops based on the src and
+      // mask contiguity/alignment to avoid rebuilding ModuleAxisInfoAnalysis
+      // after every IR change.
+      triton::ModuleAxisInfoAnalysis axisAnalysis(m);
+      DenseMap<ttg::AsyncCopyGlobalToLocalOp, unsigned> asyncCopyContiguity;
+      m->walk([&](ttg::AsyncCopyGlobalToLocalOp copyOp) {
+        unsigned contiguity =
+            mlir::LLVM::AMD::getContiguity(copyOp.getSrc(), axisAnalysis);
+        if (auto mask = copyOp.getMask()) {
+          contiguity = std::min<unsigned>(contiguity,
+                                          axisAnalysis.getMaskAlignment(mask));
+        }
+        asyncCopyContiguity.insert({copyOp, contiguity});
+      });
+      patterns.add<CoalesceAsyncCopyWrites>(targetInfo, asyncCopyContiguity,
+                                            context);
+      break;
+    }
+    default:
+      break;
```

Comment:
I think we can change this to `return;` and so that we can move L162 - L177 to after the `switch` statement.